File: leetcode/findMinArrowShots.js
// https://leetcode.com/problems/minimum-number-of-arrows-to-burst-balloons/

var findMinArrowShots = function (points) {
  points.sort((a, b) => a[1] - b[1]);

  //! solution 1
  //   let num = 1,
  //     previous = 0;
  //   for (let current = 1; current < points.length; current++) {
  //     // if starting of current balloon is after the ending of previous  alloon, then you'll need a new arrow to burst the current balloon
  //     if (points[current][0] > points[previous][1]) {
  //       num++;
  //       previous = current;
  //     }
  //   }
  //   return num;

  //! solution 2
  let num = 1,
    curEnd = points[0][1],
    start,
    end;
  for (let i = 0; i < points.length; i++) {
    (start = points[i][0]), (end = points[i][1]);
    if (start > curEnd) {
      num++;
      curEnd = end;
    }
  }
  return num;
};

let points =
  // [
  //   [1, 2],
  //   [3, 4],
  //   [5, 6],
  //   [7, 8],
  // ];
  [
    [10, 16],
    [2, 8],
    [1, 6],
    [7, 12],
  ];
let res = findMinArrowShots(points);
console.log(res);

File: leetcode/findMinArrowShots.py
# https://leetcode.com/problems/minimum-number-of-arrows-to-burst-balloons/?envType=daily-question&envId=2024-03-18

from typing import List

# decrement num when new overlapping interval is found
def findMinArrowShots(points: List[List[int]]) -> int:
    points.sort()
    num = len(points) # worstcase, you'll need at least n arrows
    prev = points[0]

    for i in range(1, len(points)):
        curr = points[i]

        if curr[0] <= prev[1]: #overlapping intervals!
            num -= 1
            prev = [curr[0], min(curr[1], prev[1])]
        else:
            prev = curr

    return num

# increment num (and interval end) when new non-overlapping interval is encountered
def findMinArrowShots(points: List[List[int]]) -> int:
    points.sort(key=lambda x:x[1]) #note the sort based on end
    num = 1
    _, currEnd = points[0]

    for i in range(1, len(points)):
        start, end = points[i]

        if start > currEnd:
            num += 1
            currEnd = end

    return num

points = [[9,12],[1,10],[4,11],[8,12],[3,9],[6,9],[6,7]]
print(findMinArrowShots(points))
print(findMinArrowShots(points))